### ISSUE TITLE
Crash after IPC connection failure

### ISSUE THREAD

**Issue by frozeman** [atom/electron#2064]
I created a connection to a IPC file on mac using the node native `net` module. When the connection creation fails due to `NSURLErrorDomain -1004` (Note that the errors before are logged from the net.on('error') event).

This happens when i try to open two connections over the same IPC socket file.

```
ERROR { [Error: This socket has been ended by the other party] code: 'EPIPE' } Error: This socket has been ended by the other party
ERROR { [Error: This socket has been ended by the other party] code: 'EPIPE' } Error: This socket has been ended by the other party
ERROR { [Error: This socket has been ended by the other party] code: 'EPIPE' } Error: This socket has been ended by the other party
ERROR { [Error: This socket has been ended by the other party] code: 'EPIPE' } Error: This socket has been ended by the other party
[0626/111521:ERROR:http_transport_mac.mm(186)] Could not connect to the server. (NSURLErrorDomain -1004)
```

Electron crashes with the following log: https://gist.github.com/frozeman/8eadacca0750a0785b8f

**Comment by frozeman** [comment]
I figured that it only crashes when i use an incoming sendSync IPC connection, to forward to the `net` socket.

A simplified example looks like this:
```js
var syncEvents = {};

// create connection on "ipcSocket"

// wait for data on the socket
this.ipcSocket.on("data", function(data){
    // SEND SYNC back
    _this.syncEvents['xyz'].returnValue = data.toString();
});

ipc.on('ipcProvider-writeSync', function(event, payload){
    ipcSocket.write(payload);
    syncEvents['xyz'] = event;
});
```

If you do the same async, it works fine.
Hope that helps.

**Comment by zcbenz** [comment]
Can you provide a demo app? I'm unable to reproduce according to your descriptions.

**Comment by frozeman** [comment]
I figured out what causes the issue, and also created a repro app here: https://github.com/frozeman/electron-connect-crash

This one seems to be related to my other issue: https://github.com/atom/electron/issues/2031

What happens is that i try to re-connect when the net error event is fired, as it could be the the socket was closed from the otherside and i want to reconnect.

The demo app itself, creates a web view, which then creates a IPC connection to the main process and starts a IPC connection using the `net` package to whatever IPC socket (doesn't need to work, to test the crash)

But because the web view, or window is closed it seems that the connect failure of the net package kills electron.

**To try for yourself:**
Run the demo app and then close the window (tested on mac osx).
It will then after ~1 sec kill electron.

The main line here is `/ipcProvidorBackend.js:86

Ideally this error should be handled in a way that it doesn't cause electron to crash. And stop the net processes automatically, if possible.
